File: demo/python/demo_001.py
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import List, Optional
import httpx

# Enum Class
class UserError(Enum):
    ERROR = "ERROR"
    INVALID_INPUT_TYPES = "INVALID_INPUT_TYPES"

# Custom Exception Class
class CustomException(Exception):
    def __init__(self, error: UserError, message: str = ""):
        self.error = error
        self.message = message
        super().__init__(f"{error.value}: {message}")

# DataClass
@dataclass(frozen=True)
class User:
    id: int
    name: str
    birthday: datetime

    def get_age(self) -> int:
        return int((datetime.now() - self.birthday).days / 365.25)

# ServiceClass
class UserService:
    _url: str
    _users: List[User] = []

    @property
    def url(self) -> str:
        return self._url

    def __init__(self, url: str):
        self._url = url

    async def fetch(self) -> bool:
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(self._url)
                response.raise_for_status()
                users_data = response.json()
                self._users = self._decode(users_data=users_data)
                return True
            except Exception as e:
                raise CustomException(UserError.ERROR, str(e)) from e

    def _decode(self, users_data: List[dict]) -> List[User]:
        _list = []
        for user_data in users_data:
            try:
                id = user_data['id']
                name = user_data['name']
                birthday = user_data['birthday']
                if not (isinstance(id, int) and isinstance(name, str) and isinstance(birthday, str)):
                    raise CustomException(UserError.INVALID_INPUT_TYPES)
                birthday = datetime.strptime(birthday, "%Y-%m-%d")
                _list.append(User(id=id, name=name, birthday=birthday))
            except KeyError as e:
                raise CustomException(UserError.INVALID_INPUT_TYPES, f"Missing key: {e}") from e
        return _list

    def get_user_by_index(self, index: int) -> Optional[User]:
        try:
            return self._users[index]
        except IndexError:
            return None

    def get_user_by_id(self, id: int) -> Optional[User]:
        for user in self._users:
            if user.id == id:
                return user
        return None


# ------------------------------------


import asyncio

api_url_001 = "http://localhost:4010/users"
api_url_002 = "http://localhost:4010/"

async def main():

    print("--- test001 ---")

    users = UserService(api_url_001)
    print(f"url: {users.url}")
    try:
        await users.fetch()
        user = users.get_user_by_index(0)
        print(f"user: {user}")
        print(f"age: {user.get_age()}")
        user = users.get_user_by_id(2)
        #user.id = 100
        #print(f"id: {user.id}")
        print(f"user: {user}")
        print(f"age: {user.get_age()}")
    except CustomException as error:
        print("error", error)

    print("--- test002 ---")

    users = UserService(api_url_002)
    print(f"url: {users.url}")
    try:
        await users.fetch()
    except CustomException as error:
        print("error", error)

asyncio.run(main())

File: demo/js/demo_001.js
'use strict';

// Enum
const UserError = Object.freeze({
  ERROR: "ERROR",
  INVALID_INPUT_TYPES: "INVALID_INPUT_TYPES",
});

// DataClass
class User {
  constructor(id, name, birthday) {
    this.id = id;
    this.name = name;
    this.birthday = new Date(birthday);
  }

  getAge() {
    const today = new Date();
    const birthDate = new Date(this.birthday);
    let age = today.getFullYear() - birthDate.getFullYear();
    const month = today.getMonth() - birthDate.getMonth();
    if (month < 0 || (month === 0 && today.getDate() < birthDate.getDate())) {
      age--;
    }
    return age;
  }
}

// ServiceClass
class UserService {
  #users;

  constructor(url) {
    if (typeof url !== "string") {
      throw new Error(UserError.INVALID_INPUT_TYPES);
    }

    Object.defineProperty(this, "url", { value: url, writable: false });
    this.#users = [];
  }

  async fetch() {
    let jsonData;
    try {
      const response = await fetch(this.url);
      jsonData = await response.json();
    } catch (error) {
      throw new Error(UserError.ERROR);
    }

    if (!Array.isArray(jsonData)) {
      throw new Error(UserError.ERROR);
    }

    this.#users = this.#decode(jsonData);

    return true;
  }

  #decode(jsonData) {
    return jsonData.map((user) =>
      this.#createUser(user.id, user.name, user.birthday)
    );
  }

  #createUser(id, name, birthday) {
    if (
      typeof id !== "number" ||
      typeof name !== "string" ||
      typeof birthday !== "string"
    ) {
      throw new Error(UserError.INVALID_INPUT_TYPES);
    }

    const birthdayDate = new Date(birthday);
    if (isNaN(birthdayDate.getTime())) {
      throw new Error(UserError.INVALID_INPUT_TYPES);
    }
    return Object.freeze(new User(id, name, birthdayDate));
  }

  getUserByIndex(index) {
    if (index < 0 || index >= this.#users.length) {
      return null;
    }
    return this.#users[index];
  }

  getUserById(id) {
    return this.#users.find(user => user.id === id) || null;
  }
}

// ----------------------------------------

const api_url_001 = "http://localhost:4010/users";
const api_url_002 = "http://localhost:4010/";

async function main() {
  console.log("--- test001 ---");

  const usersService1 = new UserService(api_url_001);
  console.log(`url: ${usersService1.url}`);
  try {
    await usersService1.fetch();
    console.log(`private var: ${usersService1.users}`) // undefined
    let user = usersService1.getUserByIndex(0);
    console.log(`user: ${JSON.stringify(user)}`);
    console.log(`age: ${user.getAge()}`);
    user = usersService1.getUserById(2);
    //user.id = 100;
    //console.log(`id: ${user.id}`);
    console.log(`user: ${JSON.stringify(user)}`);
    console.log(`age: ${user.getAge()}`);
  } catch (error) {
    console.error("error", error);
  }

  console.log("--- test002 ---");

  const usersService2 = new UserService(api_url_002);
  console.log(`url: ${usersService2.url}`);

  // Error
  try {
    await usersService2.fetch();
  } catch (error) {
    console.error("error", error);
  }
}

main();
